# Snakefile 

# Prerequisites (on Gadi): source /g/data/pq84/malaria/snakemake_pipeline/snakemake_setup.sh

# Define paths for ref genomes using config file
configfile: "config/config.yaml"
fasta_path=config['fasta_path']
picard_path=config['picard_path']
gatk_path=config['gatk_path']
known_indels_path=config['known_indels_path']
known_sites_path=config['known_sites_path']
bed_file_path=config['bed_file_path']
source_dir=config['source_dir']

# Define input files - 

## Defining the samples to be used for the {sample} wildcard

SAMPLES, = glob_wildcards(f'{source_dir}/{{sample}}_1.fastq.gz') 

## Chromosome names for bed files and subsetting jobs

with open(config['bed_file_path']) as f:
    CHROMOSOME = f.read().splitlines()
    CHROMOSOME = [p.split('\t')[0] for p in CHROMOSOME]

# Create bed for each chromosome/contig - to be used for sub-setting jobs (eg joint calling)

with open(config['bed_file_path']) as f:
    chrom_bed = f.read().splitlines()
    chrom_bed = [p.split('\t') for p in chrom_bed]

for i in chrom_bed:
    content = '\t'.join(i)
    file_name = i[0]
    file_path = 'data/chromosomes/' + file_name + '.bed'
    f = open(file_path, 'w') # amend this line - how to amend bed file path and use chrom as file name
    f.write(content)

f.close()

# Define final files

rule all:
    input:
        expand("output/mapped_reads/{sample}.bam", sample = SAMPLES),
        expand("output/sorted/{sample}_sorted.bam", sample = SAMPLES),
        expand("output/sorted/{sample}_sorted.bam.bai", sample = SAMPLES),
        expand("output/bam_recal/{sample}_dupmarked.bam", sample = SAMPLES),
        expand("output/bam_recal/{sample}_picard_metrics_file.txt", sample = SAMPLES),
        expand("output/bam_recal/{sample}_dupmarked_reheader.bam", sample = SAMPLES),
        expand("output/bam_recal/{sample}_dupmarked_reheader.bam.bai", sample = SAMPLES),
        expand("output/calling/gatk/gvcf/{sample}.g.vcf.gz", sample = SAMPLES),
        "output/calling/gatk/gvcf/GATK_combined.g.vcf.gz",
        expand("output/calling/gatk/joint/gatk_genotyped_{chromosome}.vcf.gz", chromosome = CHROMOSOME), 
        expand("output/calling/bcftools/input_bam_files.list"),
        expand("output/calling/bcftools/bcftools_genotyped_{chromosome}.vcf.gz", chromosome = CHROMOSOME),
        expand("output/calling/bcftools/bcftools_genotyped_{chromosome}.vcf.gz.tbi", chromosome = CHROMOSOME),
        expand("output/calling/consensus/{chromosome}_consensus.vcf.gz", chromosome = CHROMOSOME),
        expand("output/calling/consensus/{chromosome}_consensus.vcf.gz.tbi", chromosome = CHROMOSOME),
        expand("output/calling/consensus/{chromosome}.txt", chromosome = CHROMOSOME),
        "output/calling/consensus/Consensus.vcf.gz",
        "output/calling/consensus/Consensus.vcf.gz.tbi"

# Define local rules - not run with scheduler
localrules: all, bam_input_list

# Concatenate chromosome-based consensus VCFs 
rule concat_vcfs:
    params:
        vcf = lambda w: " ".join(expand("output/calling/consensus/{chromosome}_consensus.vcf.gz", chromosome = CHROMOSOME))
    output:
        vcf="output/calling/consensus/Consensus.vcf.gz",
        tbi="output/calling/consensus/Consensus.vcf.gz.tbi"
    shell:
        """
        bcftools concat -o {output.vcf} {params.vcf}
        bcftools index -t -o {output.tbi} {output.vcf}
        """

# Take a consenus of GATK and bcftools
rule consensus_of_vcfs:
    input:
        bcftools="output/calling/bcftools/bcftools_genotyped_{chromosome}.vcf.gz",
        gatk="output/calling/gatk/joint/gatk_genotyped_{chromosome}.vcf.gz"
    output:
        txt="output/calling/consensus/{chromosome}.txt",
        vcf="output/calling/consensus/{chromosome}_consensus.vcf.gz",
        tbi="output/calling/consensus/{chromosome}_consensus.vcf.gz.tbi"
    params:
        header = lambda w: f"'%CHROM\\t%POS\\n'"
    shell:
        """
        bcftools query -f {params.header} {input.bcftools} > {output.txt}
        bcftools filter -R {output.txt} -o {output.vcf} {input.gatk}
        bcftools index -t -o {output.tbi} {output.vcf}
        """

# Run bcftools for each chromosome
rule bcftools_caller:
    input:
        input_bam_files="output/calling/bcftools/input_bam_files.list",
        fasta=fasta_path
    output:
        vcf="output/calling/bcftools/bcftools_genotyped_{chromosome}.vcf.gz",
        tbi="output/calling/bcftools/bcftools_genotyped_{chromosome}.vcf.gz.tbi"
    shell:
        """
        bcftools mpileup --threads 2 -f {input.fasta} -b {input.input_bam_files} | bcftools call --threads 2 -m -Oz -a FORMAT/GQ,FORMAT/GP,INFO/PV4 -v -o {output.vcf}
        bcftools index --threads 2 -t -o {output.tbi} {output.vcf}
        """

# Create input list of bam files for bcftools

rule bam_input_list:
    output:
        temp("output/calling/bcftools/input_bam_files.list")
    run:
        import glob

        bam_list = glob.glob('output/bam_recal/*_dupmarked_reheader.bam')
        #bam_list = [sub.replace('output/bam_recal/', '') for sub in bam_list] 

        file = open('output/calling/bcftools/input_bam_files.list', 'w')
        for item in bam_list:
            file.write(item+"\n")

        file.close()

# Joint-call variants

rule joint_genotyping:
    input:
        vcf="output/calling/gatk/gvcf/GATK_combined.g.vcf.gz",
        fasta=fasta_path,
        bed="data/chromosomes/{chromosome}.bed"
    output:
        "output/calling/gatk/joint/gatk_genotyped_{chromosome}.vcf.gz"
    params:
        gatk=gatk_path
    shell:
        """
        java -Djava.iodir=1000m -Xms3200m -Xmx3600m -jar {params.gatk} \
        -T GenotypeGVCFs \
        -nt 3 \
        -R {input.fasta} \
        -L {input.bed} \
        -V {input.vcf} \
        -o {output}
        """
            
# Combine gVCFs

rule combine_gvcfs:
    input:
        fasta=fasta_path
    output:
        "output/calling/gatk/gvcf/GATK_combined.g.vcf.gz"
    params:
        gatk=gatk_path,
        gvcfs = lambda w: " -V " + " -V ".join(expand("output/calling/gatk/gvcf/{sample}.g.vcf.gz", sample = SAMPLES))
    shell:
        """
        java -Djava.iodir=1000m -Xms3200m -Xmx3600m -jar {params.gatk} \
        -T CombineGVCFs \
        -R {input.fasta} \
        {params.gvcfs} \
        -o {output}
        """

# Call haplotypes - GATK

rule haplotype_caller:
    input:
        bam="output/bam_recal/{sample}_dupmarked_reheader.bam",
        fasta=fasta_path
    output:
        "output/calling/gatk/gvcf/{sample}.g.vcf.gz"
    params:
        gatk=gatk_path
    shell:
        """
        java -Djava.iodir=1000m -Xms3200m -Xmx3600m -jar {params.gatk} \
        -T HaplotypeCaller \
        -ERC GVCF \
        --minPruning 3 \
        --maxNumHaplotypesInPopulation 200 \
        --max_alternate_alleles 3 \
        --variant_index_type LINEAR \
        --variant_index_parameter 128000 \
        -contamination 0.0 \
        -G Standard \
        -R {input.fasta} \
        -I {input.bam} \
        -o {output}
        """

# Update headers and index bam files

rule update_header_and_index:
    input:
        "output/bam_recal/{sample}_dupmarked.bam"
    output:
        bam_output=temp("output/bam_recal/{sample}_dupmarked_reheader.bam"),
        bam_index=temp("output/bam_recal/{sample}_dupmarked_reheader.bam.bai")
    params:
        header = lambda w: f"'s,^@RG.*,@RG\\tID:{w.sample}\\tSM:{w.sample}\\tLB:None\\tPL:Illumina,g'"
    shell:
        """
        samtools view -H {input} | \
        sed {params.header} | \
        samtools reheader - {input} > {output.bam_output}

        samtools index {output.bam_output} 
        """

# Mark duplicates

rule mark_duplicates:
    input: 
        bam="output/sorted/{sample}_sorted.bam",
        bam_index="output/sorted/{sample}_sorted.bam.bai"
    output: 
        dup_marked=temp("output/bam_recal/{sample}_dupmarked.bam"),
        metrics_file=temp("output/bam_recal/{sample}_picard_metrics_file.txt")
    params:
        picard=picard_path
    shell:
        """
        java -Djava.iodir=1000m -jar {params.picard} \
        MarkDuplicates AS=TRUE VALIDATION_STRINGENCY=LENIENT \
        I={input.bam} \
        O={output.dup_marked} \
        M={output.metrics_file}
        """
        
# Sort BAM for recalibration & index

rule index_sorted_bam: 
    input: 
        "output/sorted/{sample}_sorted.bam"
    output:
        temp("output/sorted/{sample}_sorted.bam.bai") 
    shell:
        "samtools index {input}"

rule samtools_sort:
    threads: 5
    input: 
        "output/mapped_reads/{sample}.bam"
    output:
        temp("output/sorted/{sample}_sorted.bam")
    shell:
        "samtools sort -@ {threads} {input} > {output}"

# Map reads 

rule bwa_map:
    threads: 5
    input:
        fasta_path,
        f"{source_dir}/{{sample}}_1.fastq.gz",
        f"{source_dir}/{{sample}}_2.fastq.gz"
    output:
        "output/mapped_reads/{sample}.bam"
    params:
        header = lambda w: f"@RG\\\\tID:{w.sample}\\\\tPL:ILLUMINA"
    shell:
         "bwa mem -t {threads} -M -R {params.header} {input} | samtools view -u -S - | samtools sort -n -o {output}"
